### PR TITLE
Release v2.0.0

### PR DESCRIPTION
**Pull request by sk-sahu**
Description - https://github.com/lifebit-ai/exomiser-pipeline-nf/pull/3#issuecomment-1353362425

### CODE REVIEW

**Comment by l-mansouri** [comment]
## Overview
This PR modifies the behaviour of the pipeline to take as input multi-sample VCF files.

## Purpose

This PR allows the exomiser pipeline to accept multiple samples in the form of a design file. 

The design file is a tab separated file that contains the location of the VCF files and family information. The family information is used by the first process to create 3 files( ID file (contains proband ID for VCF);  PED file (contains family information for samples within the multi sample VCF); HPO file (the phenotype IDs the user is interested in))

The design file can also be used to run the pipeline on a single-sample VCF simply using `nan` in the columns corresponding to the parent ids.

## Acceptance Criteria/Jira Ticket


[DEL-15823](https://lifebit.atlassian.net/browse/DEL-15823)

## Changes

- Implements families_file parameter (for the design file)
- Changes the input (now uploaded to S3 bucket)
- Adds container for the production of the PED file
- Adds process that accepts design file
- Adds new testing profiles to test the two behaviour of the code ( `single_vcf_test`, `family_test`)
- Cleans the code
- Fixes bugs

## Internal CloudOS Tests

- [ ] [-profile < family_test >](https://cloudos.lifebit.ai/app/jobs/6540e0c3629e3a9a787f6aa9)
- [ ] [-profile < single_vcf_test >](https://cloudos.lifebit.ai/app/jobs/6540d5db629e3a9a787ecb73)

## Client CloudOS Tests

- [ ] [-profile <replace_with_profile1>]()
- [ ] [-profile <replace_with_profile2>]()

## PR Checklist

#### Author to check:

- [x] The PR title is informative and begins with appropriate `[Fix:|Feat:|Docs:]` based on PR type
- [ ] This PR is raised to `dev` branch for DSL1 or `DSL2-dev` for DSL2 translations
- [x] The description contains the acceptance criteria or link to the associated Jira ticket
- [ ] The description contains links to CloudOS tests covering all test cases in client environment

#### Reviewer to check:

- [ ] A new test profile covering new functions for acceptance criteria is created if it does not already exist
- [ ] Nextflow configs have associated profiles named to match the config basename
- [ ] The [Jenkinsfile](https://github.com/lifebit-ai/<target_repo_name>/blob/dev/Jenkinsfile) has been updated with current Nextflow profiles
- [ ] The [internal_lifebit_cloudos_ci.yml](https://github.com/lifebit-ai/<target_repo_name>/blob/dev/.github/workflows/internal_lifebit_cloudos_ci.yml) has been updated to run all current Nextflow profiles
- [ ] Redundant tests have been removed
- [ ] The [README.md](https://github.com/lifebit-ai/<target_repo_name>/blob/dev/docs/README.md) has been updated with parameter changes and current `results/` and conforms to the README.md guidelines
- [ ] The code is logical, readable, reasonably optimal and meets the acceptance criteria
- [ ] The [containers](https://github.com/lifebit-ai/<target_repo_name>/blob/dev/containers) folder has been updated with the necessary containers to run this pipeline if needed

File: .attach_pid22597
Comment (sk-sahu) [inline .attach_pid22597:1]: This file can be removed.

File: bin/add_exomiser_fields_to_genotiers.js
Comment (sk-sahu) [inline bin/add_exomiser_fields_to_genotiers.js:1]: Let's remove this file and add this to `.gitignore`

File: README.md [hunk 1]
```diff
@@ -1,54 +1,110 @@
...
+nextflow run main.nf -profile family_test
 ```
 
-### Test run (With a HOP term file, no fetch from database)
+To test the pipeline on a single-sample VCF:
 
-```bash
-nextflow run main.nf -profile test
 ```
+nextflow run main.nf -profile single_vcf_test
```
Comment (sk-sahu) [inline README.md:95]: A command line version with all the test parameters and their values

File: bin/genotiers.js
Comment (sk-sahu) [inline bin/genotiers.js:1]: This file also can be removed

File: bin/get_hpo_terms_from_barcode.js
Comment (sk-sahu) [inline bin/get_hpo_terms_from_barcode.js:1]: This file also can be removed

File: main.nf [hunk 2]
```diff
@@ -46,19 +44,30 @@ log.info ""
...
-if(params.input) {
-   Channel
-      .fromPath( "${params.input}" )
-      .ifEmpty { exit 1, "VCF file: ${params.input} not found"}
-      .into { ch_vcf ; ch_vcf_inspect; ch_vcf_for_geneyx }
+if(params.families_file) {
+  Channel
+      .fromPath( "${params.families_file}")
+      .ifEmpty { exit 1, "Family file: ${params.families_file} not found"}
+      .set {ch_vcf}
```
Comment (sk-sahu) [inline main.nf:51]:
```suggestion
      .set {ch_families_file}
```

File: main.nf [hunk 3]
```diff
@@ -46,19 +44,30 @@ log.info ""
...
 }
 
+
+
+Channel
+    .fromPath(params.families_file)
+    .ifEmpty { exit 1, "Cannot find input file : ${params.families_file}" }
+    .splitCsv(header:true, sep:'\t', strip: true)
+    .map {row -> [ row.proband_id, file(row.vcf_path), file(row.vcf_index_path)] }
+    .set {ch_input1}
```
Comment (sk-sahu) [inline main.nf:63]:
```suggestion
    .set {ch_vcf_paths}
```

File: testdata/fam_file.tsv [hunk 4]
```diff
@@ -0,0 +1,2 @@
+run_id	proband_id	hpo	vcf_path	vcf_index_path	proband_sex	mother_id	father_id
+EX001	ERR3239334	HP:0001156	s3://lifebit-featured-datasets/pipelines/exomiser-nf/family_ERR3239334_ERR3989341_ERR3989342_small.vcf.gz	s3://lifebit-featured-datasets/pipelines/exomiser-nf/family_ERR3239334_ERR3989341_ERR3989342_small.vcf.gz.tbi	M	ERR3989342	ERR3989341
```
Comment (sk-sahu) [inline testdata/fam_file.tsv:2]: Needs to check if it's possible to support multiple HOP terms using this input format, similar to v1 of this pipeline.

File: main.nf [hunk 5]
```diff
@@ -78,111 +87,122 @@ println(selected_prioritisers)
...
-  tag "${vcf}-${prioritiser}"
-  publishDir "${params.outdir}/${sample_name}", mode: 'copy'
 
+process exomiser {
+  tag "${vcf_path1}"
+  publishDir "${params.outdir}/${proband_id1}", mode: 'copy'
+  maxForks 1
+  submitRateLimit = '1 / 5 m'
+  errorStrategy 'retry'
+  maxRetries 3
```
Comment (sk-sahu) [inline main.nf:144]: This can be parameterised and defaults can be in config file

File: main.nf [hunk 6]
```diff
@@ -78,111 +87,122 @@ println(selected_prioritisers)
...
-  file(vcf) from ch_vcf
-  //The following is expected when CADD is omitted,
-  // WARN: Input tuple does not match input set cardinality declared by process `exomiser`
-  // ch_all_exomiser_data contents can be 1 or 2 folders, (exomiser_data +/- cadd separately)
-  // this is fine, as when there is no second dir, a fake input.1 is generated that will be unused
-  file(application_properties) from ch_application_properties
-  file(auto_config_yml) from ch_auto_config_yml
-  file(exomiser_data) from ch_exomiser_data
-  val(hpo_terms) from ch_hpo_terms
+  set val(proband_id1),file(vcf_path1),file(vcf_index1), file(hpo_file), file(ped_file),file(id_file) from combined_channel
```
Comment (sk-sahu) [inline main.nf:146]:
```suggestion
  set val(proband_id1),file(vcf_path1),file(vcf_index1), file(hpo_file), file(ped_file),file(id_file) from ch_combined
```

File: main.nf [hunk 7]
```diff
@@ -78,111 +87,122 @@ println(selected_prioritisers)
...
+    # Run Exomiser
+    ${exomiser} \
+    --analysis new_auto_config.yml \
+    --spring.config.location=$application_properties \
+    --exomiser.data-directory='.'
+    # Create the slot for CloudOS html report preview
+    mkdir MultiQC
+    cp *.html MultiQC/multiqc_report.html
+
+    sed -i  "s/Anonymous/\$proband_id1/" MultiQC/multiqc_report.html
```
Comment (sk-sahu) [inline main.nf:200]: Modify the code to have the `MultiQC/multiqc_report.html` on base `results` folder

File: nextflow.config [hunk 8]
```diff
@@ -54,21 +57,23 @@ params {
...
-    standard { includeConfig params.config }
-    awsbatch { includeConfig 'conf/executors/awsbatch.config' }
-    eu_west_1 { includeConfig 'conf/cloud-region/eu_west_1.config' }
-    eu_west_2 { includeConfig 'conf/cloud-region/eu_west_2.config' }
-    test_full { includeConfig "conf/tests/full/test_full.config" }
-    ci_test_data { includeConfig "conf/data/ci_test_data.config" }
-    quay { includeConfig 'conf/containers/quay.config' }
+   standard { includeConfig params.config }
+   family_test { includeConfig 'conf/family_test.config' }
+   single_vcf_test { includeConfig 'conf/single_vcf_test.config' }
```
Comment (sk-sahu) [inline nextflow.config:62]:
```suggestion
   test_family { includeConfig 'conf/family_test.config' }
   test_single_vcf { includeConfig 'conf/single_vcf_test.config' }
```

File: bin/genotiers.js
Comment (l-mansouri) [reply bin/genotiers.js:1]: this file, has some code and it's called by exomiser!

File: bin/get_hpo_terms_from_barcode.js
Comment (l-mansouri) [reply bin/get_hpo_terms_from_barcode.js:1]: this file, has some code and it's called by exomiser!

File: testdata/fam_file.tsv (hunk 4 above)
Comment (l-mansouri) [reply testdata/fam_file.tsv:2]: this has been tested now!

**Review by l-mansouri — APPROVED** [review]

**Pull request merged**